Fix race condition in testVMManagerStopTheWorld().

MenloDorian opened the pull request:
#### 24181825972e64b0adf9bc4250b28cdac127ccce
<pre>
Fix race condition in testVMManagerStopTheWorld().
<a href="https://bugs.webkit.org/show_bug.cgi?id=300335">https://bugs.webkit.org/show_bug.cgi?id=300335</a>
<a href="https://rdar.apple.com/162135022">rdar://162135022</a>

Reviewed by Yijia Huang.

The global totalNumberOfVMs is being incremented from worker threads as they start.  As such, there&apos;s
a race where their read-modify-write operation can trash each other&apos;s mods.  The fix is simply to make
totalNumberOfVMs atomic.

Also made the following changes:

1. Added comments for each case in checkpointCallback() to make it clear that these represent the
   various checkpoints in the test.  Since the switch statement is large, it is not alwats easy
   to realize that this is a switch on the checkpoints without having to scroll up and down a lot.
   The comments makes it easier to confirm that we&apos;re reading checkpoint code.

2. Flush stdout and stderr at the top of VMManagerStopTheWorldTest::test().
   Previous test runs before running this one has printed a lot of messages to stdout.
   VMManagerStopTheWorldTest uses WTF::dataLog(), which prints to stderr.  Forcing all previous
   messages to be flushed will ensure that VMManagerStopTheWorldTest&apos;s error messages don&apos;t
   show up in the middle of the stdout of preceding tests.  This is a qualifty of life improvement
   for diagnosing test failures in here.

3. Make the ABORT_IF_FAILED() macro take the locker argument to make it clear that it requires
   the test lock to already be held.  Previously, this would take a compiler error to reveal this
   requirement.  Making it explicit improved clarity at a glance.

Test: Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp
* Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp:
(VMManagerStopTheWorldTest::checkpointCallback):
(VMManagerStopTheWorldTest::test):

Canonical link: <a href="https://commits.webkit.org/301159@main">https://commits.webkit.org/301159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ce39d25476e44135417fd40e467bbe03ec27732

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125118 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44785 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35524 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131966 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fe8a50bd-52cd-420d-be09-14f8e5ce05e3) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45476 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53348 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/95247 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/89faf88a-106a-4c1f-b087-efd6d7915466) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128072 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36302 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/111887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75789 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/afc2d5bc-c07b-4362-bde4-d1db922ccfb9) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/35208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30050 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75446 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117209 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106067 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30280 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134646 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123636 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51928 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39720 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103713 "39 flakes 67 failures") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52362 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108108 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103484 "Found 1 new API test failure: WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/service-worker-registrations (failure)") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48827 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/27121 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19598 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51820 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/57604 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156659 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51187 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39234 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54543 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52877 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1ce39d25476e44135417fd40e467bbe03ec27732. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51958

hyjorc1 reviewed (approved):
r=me

hyjorc1 commented on Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp:
```diff
@@ -222,7 +222,7 @@ Vector<VM*>* testVMsPtr = nullptr;
 
-unsigned totalNumberOfVMs = 0;
+Atomic<unsigned> totalNumberOfVMs = 0;
 unsigned totalNumberOfActiveVMs = 0;
```
Do we need atomic for `totalNumberOfActiveVMs` also? Have you check all the other number variables in this test? Do they need atomic?

MenloDorian replied:
`totalNumberOfActiveVMs` is fine.  It is only modified by the main thread except for one place in the worker, and the main thread is waiting at that point while the worker runs.  As for the others, they also look fine based on my previous audit of them.

Commit:
Fix race condition in testVMManagerStopTheWorld().
https://bugs.webkit.org/show_bug.cgi?id=300335
rdar://162135022

Reviewed by Yijia Huang.

The global totalNumberOfVMs is being incremented from worker threads as they start.  As such, there's
a race where their read-modify-write operation can trash each other's mods.  The fix is simply to make
totalNumberOfVMs atomic.

Also made the following changes:

1. Added comments for each case in checkpointCallback() to make it clear that these represent the
   various checkpoints in the test.  Since the switch statement is large, it is not alwats easy
   to realize that this is a switch on the checkpoints without having to scroll up and down a lot.
   The comments makes it easier to confirm that we're reading checkpoint code.

2. Flush stdout and stderr at the top of VMManagerStopTheWorldTest::test().
   Previous test runs before running this one has printed a lot of messages to stdout.
   VMManagerStopTheWorldTest uses WTF::dataLog(), which prints to stderr.  Forcing all previous
   messages to be flushed will ensure that VMManagerStopTheWorldTest's error messages don't
   show up in the middle of the stdout of preceding tests.  This is a qualifty of life improvement
   for diagnosing test failures in here.

3. Make the ABORT_IF_FAILED() macro take the locker argument to make it clear that it requires
   the test lock to already be held.  Previously, this would take a compiler error to reveal this
   requirement.  Making it explicit improved clarity at a glance.

Test: Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp
* Source/JavaScriptCore/API/tests/VMManagerStopTheWorldTest.cpp:
(VMManagerStopTheWorldTest::checkpointCallback):
(VMManagerStopTheWorldTest::test):

Canonical link: https://commits.webkit.org/301159@main
